ReactAndroid/src/main/java/com/facebook/react/animated/NativeAnimatedModule.java

```diff
@@ -0,0 +1,265 @@
+ * few types of those, e.g., spring, timing, decay). Animation objects can be "started" and
+ * "stopped". Those are like "pulse generators" for the rest of the nodes graph. Those pulses then
+ * propagate along the graph to the children nodes up to the special node type: AnimatedProps which
+ * then can be used to calculate property update map for a view.
+ *
+ * This class acts as a proxy between the "native" API that can be called from JS and the main class
+ * that coordinates all the action: {@link NativeAnimatedNodesManager}. Since all the methods from
+ * {@link NativeAnimatedNodesManager} need to be called from the UI thread, we we create a queue of
+ * animated graph operations that is then enqueued to be executed in the UI Thread at the end of the
+ * batch of JS->native calls (similarily to how it's handled in {@link UIManagerModule}). This
+ * isolates us from the problems that may be caused by concurrent updates of animated graph while UI
+ * thread is "executing" the animation loop.
+ */
+public class NativeAnimatedModule extends ReactContextBaseJavaModule implements
+    OnBatchCompleteListener, LifecycleEventListener {
+
+  private interface UIThreadOperation {
+    void execute(NativeAnimatedNodesManager animatedNodesManager);
+  }
+
+  private final Object mOperationsCopyLock = new Object();
+  private @Nullable GuardedChoreographerFrameCallback mAnimatedFrameCallback;
+  private @Nullable ReactChoreographer mReactChoreographer;
+  private ArrayList<UIThreadOperation> mOperations = new ArrayList<>();
+  private volatile @Nullable ArrayList<UIThreadOperation> mReadyOperations = null;
+
+  // Nodes manager will only be accessed from the UI Thread
+  private final NativeAnimatedNodesManager mNodesManager = new NativeAnimatedNodesManager();
+
+  public NativeAnimatedModule(ReactApplicationContext reactContext) {
+    super(reactContext);
+  }
+
+  @Override
+  public void initialize() {
+    // Safe to acquire choreographer here, as initialize() is invoked from UI thread.
+    mReactChoreographer = ReactChoreographer.getInstance();
+
+    ReactApplicationContext reactCtx = getReactApplicationContext();
+    final UIImplementation uiImplementation =
+      reactCtx.getNativeModule(UIManagerModule.class).getUIImplementation();
+    mAnimatedFrameCallback = new GuardedChoreographerFrameCallback(reactCtx) {
+      @Override
+      protected void doFrameGuarded(final long frameTimeNanos) {
+
+        ArrayList<UIThreadOperation> operations;
+        synchronized (mOperationsCopyLock) {
+          operations = mReadyOperations;
+          mReadyOperations = null;
+        }
+
+        if (operations != null) {
+          for (int i = 0, size = operations.size(); i < size; i++) {
+            operations.get(i).execute(mNodesManager);
+          }
+        }
+
+        if (mNodesManager.hasActiveAnimations()) {
+          mNodesManager.runUpdates(uiImplementation, frameTimeNanos);
```

Comment:
Slightly prefer passing the UIImplementation as a constructor arg to the nodes manager instead.